testbook/client.py

```diff
@@ -146,18 +165,44 @@ def inject(self, code, args=None, run=True, before=None, after=None):
         code_cell = new_code_cell(lines)
         self.cells.insert(inject_idx, code_cell)
 
-        return TestbookNode(self.execute_cell(inject_idx)) if run else TestbookNode(code_cell)
+        cell = TestbookNode(self.execute_cell(inject_idx)) if run else TestbookNode(code_cell)
 
-    def value(self, name):
-        """Extract a JSON-able variable value from notebook kernel"""
+        if run and pop:
+            self.cells.pop(inject_idx)
 
-        result = self.inject(name)
+        return cell
+
+    def value(self, code, safe=True):
+        result = self.inject(code, pop=True)
         if not self._execute_result(result):
             raise TestbookError('code provided does not produce execute_result')
 
-        code = """
-        from IPython.display import JSON
-        JSON({"value" : _})
+        inject_code = """
+            from IPython.display import JSON
+            JSON({"value" : _})
         """
-        cell = self.inject(code)
-        return cell.outputs[0].data['application/json']['value']
+
+        try:
+            outputs = self.inject(inject_code, pop=True).outputs
+            return outputs[0].data['application/json']['value']
+
+        except ValueError:
```

Comment:
When we test let's make sure code error exceptions still raise:
e.g. injected code is invalid it shouldn't return a reference